src/test/Cardano/CoinSelectionSpec.hs

```diff
@@ -304,6 +341,107 @@ prop_CoinSelectionData_coverage (CoinSelectionData inps outs) = property
+
+prop_algorithm_inputsAvailable_inputsSelected
+    :: (Ord i, Show i)
+    => CoinSelectionData i o
+    -> CoinSelectionResult i o
+    -> Expectation
+prop_algorithm_inputsAvailable_inputsSelected =
+    \(CoinSelectionData inputsAvailable _) ->
+    \(CoinSelectionResult (CoinSelection inputsSelected _ _) _) ->
+        inputsSelected `shouldSatisfy` (`isSubmapOf` inputsAvailable)
+
+prop_algorithm_inputsAvailable_inputsRemaining
+    :: (Ord i, Show i)
+    => CoinSelectionData i o
+    -> CoinSelectionResult i o
+    -> Expectation
+prop_algorithm_inputsAvailable_inputsRemaining =
+    \(CoinSelectionData inputsAvailable _) ->
+    \(CoinSelectionResult _ inputsRemaining) ->
+        inputsRemaining `shouldSatisfy` (`isSubmapOf` inputsAvailable)
+
+prop_algorithm_inputsSelected_inputsRemaining
+    :: (Ord i, Show i)
+    => CoinSelectionData i o
+    -> CoinSelectionResult i o
+    -> Expectation
+prop_algorithm_inputsSelected_inputsRemaining =
+    \(CoinSelectionData _ _) ->
+    \(CoinSelectionResult (CoinSelection selected _ _) remaining) ->
+        (selected `intersection` remaining) `shouldBe` mempty
+
+prop_algorithm_inputsSelected_inputsRemaining_inputsAvailable
+    :: (Ord i, Show i)
+    => CoinSelectionData i o
+    -> CoinSelectionResult i o
+    -> Expectation
+prop_algorithm_inputsSelected_inputsRemaining_inputsAvailable =
+    \(CoinSelectionData available _) ->
+    \(CoinSelectionResult (CoinSelection selected _ _) remaining) ->
+        (selected `union` remaining) `shouldBe` available
+
+prop_algorithm_inputsSelected_outputsSelected_change
+    :: CoinSelectionData i o
+    -> CoinSelectionResult i o
+    -> Expectation
+prop_algorithm_inputsSelected_outputsSelected_change = const $
+    \(CoinSelectionResult (CoinSelection {inputs, outputs, change}) _) ->
+        coinMapValue inputs
+            `shouldBe`
+            (coinMapValue outputs `C.add` mconcat change)
+
+prop_algorithm
+    :: CoinSelectionAlgorithm i o IO
+    -> (CoinSelectionData i o -> CoinSelectionResult i o -> Expectation)
+    -> (CoinSelectionData i o)
+    -> Property
+prop_algorithm algorithm verifyExpectation params =
+    withMaxSuccess 1_000 $ monadicIO $ QC.run $ do
+        mResult <- generateResult
+        pure $ isRight mResult ==>
+            let Right result = mResult in
+            verifyExpectation params result
+  where
+    CoinSelectionData {csdInputsAvailable, csdOutputsRequested} = params
+    generateResult = runExceptT
+        $ selectCoins algorithm
+        $ CoinSelectionParameters csdInputsAvailable csdOutputsRequested
+        $ CoinSelectionLimit
+        $ const $ fromIntegral $ F.length csdInputsAvailable
+
+isSubmapOf :: Ord k => CoinMap k -> CoinMap k -> Bool
+isSubmapOf (CoinMap a) (CoinMap b) = a `Map.isSubmapOf` b
+
+intersection :: Ord k => CoinMap k -> CoinMap k -> CoinMap k
+intersection (CoinMap a) (CoinMap b) = CoinMap $ a `Map.intersection` b
+
+union :: Ord k => CoinMap k -> CoinMap k -> CoinMap k
```

Comment:
is this really needed when we have https://github.com/input-output-hk/cardano-coin-selection/blob/master/src/library/Cardano/CoinSelection.hs#L123 ?

Reply:
ah, ok. you wanted to make it explicit

Reply:
> is this really needed when we have https://github.com/input-output-hk/cardano-coin-selection/blob/master/src/library/Cardano/CoinSelection.hs#L123 ?

Strictly speaking it's not needed!

Here my motivation was to make the tests more readable. Defining `union` (arguably) makes the test more readable at the call site.

In particular, the following two lines are now directly comparable to one another:

```hs
(selected `intersection` remaining) `shouldBe` mempty
(selected `union`        remaining) `shouldBe` available
```

Whereas if we just use `<>` instead of `union`, it's no longer immediately clear that we're invoking a union operation (IMO).